project/__main__.py

```diff
@@ -0,0 +1,801 @@
+               'window_icon': IMAGE_PATH / 'window_icon.ico'
+               }
+
+
+audio_player = pyaudio.PyAudio()
+
+AUDIO_CHUNK_SIZE = 1024
+
+
+async def mainloop_coro(root):
+    '''
+    Coroutine that mimics tk.mainloop.
+    It should sleep after update such that async tasks can be completed.
+    Will run until window is closed.
+    '''
+    try:
+        while True:
+            root.update()
+            root.update_idletasks()
+            await asyncio.sleep(0)
+    except tk.TclError:
+        return
+
+
+valid_soundcodes = [filename.replace('.wav', '')
+                    for filename in os.listdir(AUDIO_PATH)
+                    if filename.endswith('.wav')]
+
+
+def get_sound_data(soundcode):
+    '''
+    Returns a tuple of (stream, sound, sound_file) for a given soundcode.
+    Soundcode should be the name of a .wav file in AUDIO_PATH without .wav.
+    '''
+
+    sound_file = open((AUDIO_PATH / '{}.wav'.format(soundcode)), 'rb')
+    sound = wave.open(sound_file, 'rb')
+    sound_format = audio_player.get_format_from_width(sound.getsampwidth())
+    stream = audio_player.open(format=sound_format,
+                               channels=sound.getnchannels(),
+                               rate=sound.getframerate(),
+                               output=True)
+    return stream, sound, sound_file
+
+
+sound_streams = {soundcode: [get_sound_data(soundcode) for i in range(10)]
+                 for soundcode in valid_soundcodes}
+
+
+def prepare_sound(soundcode):
+    '''Add a sound to the prepared sounds (makes play_sound faster)'''
+    sound_streams[soundcode].append(get_sound_data(soundcode))
+
+
+def play_sound(soundcode='tap'):
+    '''
+    Plays a sound!
+    Soundcode must be the name of a .wav file in code_jam_4/audio.
+    Soundcode must not include the .wav extension.
+    Creates:
+    *A thread to play the audio
+    *A thread to prepare the next audio stream (to minimize latency)
+    '''
+    if soundcode not in valid_soundcodes:
+        raise ValueError("Soundcode must be one of \
+                          {}".format(valid_soundcodes))
+
+    def play_and_close():
+        stream, wave_file, path_file = sound_streams[soundcode].pop()
+        data = wave_file.readframes(AUDIO_CHUNK_SIZE)
+        while data:
+            stream.write(data)
+            data = wave_file.readframes(AUDIO_CHUNK_SIZE)
+        stream.stop_stream()
+        for sound_datum_to_close in stream, wave_file, path_file:
+            sound_datum_to_close.close()
+    threading.Thread(target=play_and_close).start()
+    threading.Thread(target=prepare_sound, args=(soundcode,)).start()
+
+
+with open(WORDS_PATH, 'r') as words_file:
+    real_words = set(json.load(words_file))
+
+
+def is_word(text):
+    '''Returns whether the given word is accepted as an English word.'''
+    assert text.isalpha()
+    return text.lower() in real_words
+
+
+LETTER_SCORES = {'a': 1, 'e': 1, 'i': 1, 'u': 1, 'n': 1, 'r': 1, 'o': 1,
```

Comment:
it's a bit unusual to see a constant suddenly appear in the middle of a file, after a bunch of function definitions. 

A better option would have been to split this into multiple files so that all constants could have been at the top of the file while still being near the code where they were used.

Reply:
Huh. I've always heard that if you have a 'static' variable, you should make it a constant and put it immediately above the function it's used in.
You're right about the multiple files thing, aha, I just found it easier to read in one file, so I figured it'd be easier for anyone else who wants to read my code. I've never enjoyed having to search multiple files to see the path of logic, especially for a project this small.
Thanks for the advice \o/

Reply:
to be clear, it is good to make these constant variables, and I do agree that they should ideally be close to the function they are used in so that it's easy to find that connection between the code and the constant -- but generally speaking a file should try to group all similar things together. 

A typical file is often be formatted with groups like this:
First a group of imports, then constants, then helper functions (prefixed `_`), then regular functions, then classes, and then any module-level code.

Whenever a file grows to a size where this is no longer practical (such as if there is a huge amount of constants for all sorts of different things at the start of the file) you should either consider splitting it up into multiple files, or keeping the constants in a separate file and just importing them.